staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go

```diff
@@ -0,0 +1,310 @@
+			expSc:  nil,
+			expErr: true,
+		},
+		{
+			name:  "translate with no parameter",
+			sc:    NewStorageClass(map[string]string{}, nil),
+			expSc: NewStorageClass(map[string]string{"requestvolumefilepath": "true"}, nil),
+		},
+		{
+			name:  "translate with storagepolicyname",
+			sc:    NewStorageClass(map[string]string{"storagepolicyname": "test-policy-name"}, nil),
+			expSc: NewStorageClass(map[string]string{"storagepolicyname": "test-policy-name", "requestvolumefilepath": "true"}, nil),
+		},
+		{
+			name:  "translate with fstype",
+			sc:    NewStorageClass(map[string]string{"fstype": "ext4"}, nil),
+			expSc: NewStorageClass(map[string]string{"csi.storage.k8s.io/fstype": "ext4", "requestvolumefilepath": "true"}, nil),
+		},
+		{
+			name:  "translate with storagepolicyname and fstype",
+			sc:    NewStorageClass(map[string]string{"storagepolicyname": "test-policy-name", "fstype": "ext4"}, nil),
+			expSc: NewStorageClass(map[string]string{"storagepolicyname": "test-policy-name", "csi.storage.k8s.io/fstype": "ext4", "requestvolumefilepath": "true"}, nil),
+		},
+		{
+			name:  "translate with no parameter and allowedTopology",
+			sc:    NewStorageClass(map[string]string{}, []v1.TopologySelectorTerm{topologySelectorTerm}),
+			expSc: NewStorageClass(map[string]string{"requestvolumefilepath": "true"}, []v1.TopologySelectorTerm{topologySelectorTerm}),
+		},
+		{
+			name:  "translate with storagepolicyname and allowedTopology",
+			sc:    NewStorageClass(map[string]string{"storagepolicyname": "test-policy-name"}, []v1.TopologySelectorTerm{topologySelectorTerm}),
+			expSc: NewStorageClass(map[string]string{"storagepolicyname": "test-policy-name", "requestvolumefilepath": "true"}, []v1.TopologySelectorTerm{topologySelectorTerm}),
+		},
+		{
+			name:  "raw vSAN policy parameter should be dropped",
+			sc:    NewStorageClass(map[string]string{"hostfailurestotolerate": "test-policy-name"}, []v1.TopologySelectorTerm{topologySelectorTerm}),
+			expSc: NewStorageClass(map[string]string{"requestvolumefilepath": "true"}, []v1.TopologySelectorTerm{topologySelectorTerm}),
+		},
+	}
+	for _, tc := range cases {
+		t.Logf("Testing %v", tc.name)
+		got, err := translator.TranslateInTreeStorageClassToCSI(tc.sc)
+		if err != nil && !tc.expErr {
+			t.Errorf("Did not expect error but got: %v", err)
+		}
+
+		if err == nil && tc.expErr {
+			t.Errorf("Expected error, but did not get one.")
+		}
+
+		if !reflect.DeepEqual(got, tc.expSc) {
+			t.Errorf("Got parameters: %v, expected :%v", got, tc.expSc)
+		}
+	}
+}
+
+func TestTranslateVSphereCSIPVToInTree(t *testing.T) {
+	translator := NewvSphereCSITranslator()
+	cases := []struct {
+		name     string
+		csiPV    *v1.PersistentVolume
+		intreePV *v1.PersistentVolume
+		expErr   bool
+	}{
+		{
+			name:     "expect error when CSI PV is nil",
+			csiPV:    &v1.PersistentVolume{},
+			intreePV: nil,
+			expErr:   true,
+		},
+		{
+			name: "translate valid vSphere CSI PV to vSphere in-tree PV",
```

Comment:
Do you want to add a negative case here?

Reply:
The driver will always send a valid spec. the function `TranslateCSIPVToInTree` does return error for just two cases  pv is nil and pv.Spec.CSI  is nil.

```
	if pv == nil || pv.Spec.CSI == nil {
		return nil, fmt.Errorf("pv is nil or CSI source not defined on pv")
	}
```

Added tests for both.

```
=== RUN   TestTranslateVSphereCSIPVToInTree
--- PASS: TestTranslateVSphereCSIPVToInTree (0.00s)
    vsphere_volume_test.go:176: Testing expect error when pv is nil
    vsphere_volume_test.go:176: Testing expect error when pv.Spec.CSI is nil
    vsphere_volume_test.go:176: Testing translate valid vSphere CSI PV to vSphere in-tree PV
```